pkg/ua/ua.go

```diff
@@ -266,20 +284,34 @@ func (ua *UserAgent) handleInvite(request sip.Request, tx sip.ServerTransaction)
 	callID, ok := request.CallID()
 	if ok {
 		var transaction sip.Transaction = tx.(sip.Transaction)
-		if v, found := ua.iss.Load(*callID); found {
-			is := v.(*session.Session)
-			is.SetState(session.ReInviteReceived)
-			ua.handleInviteState(is, &request, nil, session.ReInviteReceived, &transaction)
+		branchID := utils.GetBranchID(request)
+		v, found := ua.iss.Load(NewSessionKey(*callID, branchID))
+		if toHdr, ok := request.To(); ok && toHdr.Params.Has("tag") {
+			if found {
+				is := v.(*session.Session)
+				is.SetState(session.ReInviteReceived)
+				ua.handleInviteState(is, &request, nil, session.ReInviteReceived, &transaction)
+			} else {
+				// reinvite for transaction we have no record of; reject it
+				response := sip.NewResponseFromRequest(request.MessageID(), request, sip.StatusCode(481), "Call/Transaction does not exist", "")
+				tx.Respond(response)
+			}
 		} else {
-			contactHdr, _ := request.Contact()
-			contactAddr := ua.updateContact2UAAddr(request.Transport(), contactHdr.Address)
-			contactHdr.Address = contactAddr
-
-			is := session.NewInviteSession(ua.RequestWithContext, "UAS", contactHdr, request, *callID, transaction, session.Incoming, ua.Log())
-			ua.iss.Store(*callID, is)
-			is.SetState(session.InviteReceived)
-			ua.handleInviteState(is, &request, nil, session.InviteReceived, &transaction)
-			is.SetState(session.WaitingForAnswer)
+			if found {
+				// retransmission; reject it
+				response := sip.NewResponseFromRequest(request.MessageID(), request, sip.StatusCode(482), "Loop Detected", "")
```

Comment:
im not convinced this is the correct thing to do here since, technically, you could be routing an invite back the same way it came with an altered request uri and this response would prevent that, though perhaps that is an edge case that you do not want to support

if we dont reject the call here, you would probably need to create a new invite status for re-receiving the same invite again

Reply:
I think the code is correct here. @danieldonoghue's scenario will not hit this code path as every fork must have its own branch ID (and ```found``` will be false here). Requests with the same Branch ID are either retransmits (which should have been absorbed by the transaction layer) or [merged requests](https://datatracker.ietf.org/doc/html/rfc3261#section-8.2.2.2). The above code path handles the latter, the comment should be corrected though.